// appium-ios-driver/build/lib/uiauto/uiauto-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_INSTRUMENTS_SOCKET = exports.UIAutoClient = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _uiautoResponse = _interopRequireDefault(require("./uiauto-response"));

var _logger = _interopRequireDefault(require("./logger"));

var _through = _interopRequireDefault(require("through"));

var _net = _interopRequireDefault(require("net"));

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

const MORE_COMMAND = '#more';
const DEFAULT_INSTRUMENTS_SOCKET = '/tmp/instruments_sock';
exports.DEFAULT_INSTRUMENTS_SOCKET = DEFAULT_INSTRUMENTS_SOCKET;

class UIAutoClient {
  constructor(sock = '/tmp/instruments_sock') {
    this.curCommand = null;
    this.onReceiveCommand = null;
    this.commandQueue = [];
    this.sock = sock;
    this.socketServer = null;
    this.hasConnected = false;
    this.currentSocket = null;
  }

  sendCommand(cmd) {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let cmdPromise = new _bluebird.default((resolve, reject) => {
        let cb = result => {
          if (result.status === 0) {
            resolve(result.value);
          } else if (result.status) {
            let jsonwpError = (0, _appiumBaseDriver.errorFromCode)(result.status, result.value);
            reject(jsonwpError);
          } else {
            reject(new Error(result.value));
          }
        };

        _this.commandQueue.push({
          cmd,
          cb
        });

        if (_lodash.default.isFunction(_this.onReceiveCommand)) {
          _this.onReceiveCommand();
        }
      });
      return yield cmdPromise;
    })();
  }

  start() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let connectedPromise = new _bluebird.default(resolve => {
        let response = new _uiautoResponse.default();
        _this2.socketServer = _net.default.createServer({
          allowHalfOpen: true
        }, conn => {
          if (!_this2.hasConnected) {
            _this2.hasConnected = true;

            _logger.default.info('Instruments is ready to receive commands');

            resolve(true);
          }

          conn.setEncoding('utf8');
          _this2.currentSocket = conn;
          conn.on('close', () => {
            _this2.currentSocket = null;
          });
          conn.pipe((0, _through.default)(data => {
            _logger.default.debug(`Socket data received (${data.length} bytes)`);

            response.addData(data);
          }));
          conn.on('end', () => {
            if (_this2.curCommand) {
              let result = response.getResult();

              if (result && !result.needsMoreData) {
                _this2.curCommand.cb(result);

                _this2.curCommand = null;
              } else {
                if (result) {
                  _logger.default.debug('Not the last chunk, trying to get more');
                } else {
                  _logger.default.debug('No result received. Continuing to try to get more');
                }

                _this2.commandQueue.unshift({
                  cmd: MORE_COMMAND,
                  cb: _this2.curCommand.cb
                });
              }
            } else {
              _logger.default.debug('Got a result when we were not expecting one! Ignoring it');

              response.resetBuffer();
            }

            let onReceiveCommand = () => {
              _this2.onReceiveCommand = null;
              _this2.curCommand = _this2.commandQueue.shift();

              _logger.default.debug(`Sending command to instruments: ${_this2.curCommand.cmd}`);

              conn.write(JSON.stringify({
                cmd: _this2.curCommand.cmd
              }));
              conn.end();
            };

            if (_this2.commandQueue.length) {
              onReceiveCommand();
            } else {
              _this2.onReceiveCommand = onReceiveCommand;
            }
          });
        });

        _this2.socketServer.on('close', function () {
          _logger.default.debug('Instruments socket server was closed');
        });
      });
      yield _appiumSupport.fs.rimraf(_this2.sock);
      yield (0, _appiumSupport.mkdirp)(_path.default.dirname(_this2.sock));

      _this2.socketServer.listen(_this2.sock);

      _logger.default.debug(`Instruments socket server started at ${_this2.sock}`);

      return yield connectedPromise;
    })();
  }

  shutdown() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this3.curCommand = null;
      _this3.onReceiveCommand = null;

      if (_this3.currentSocket) {
        _logger.default.debug('Destroying instruments client socket.');

        _this3.currentSocket.end();

        _this3.currentSocket.destroy();

        _this3.currentSocket = null;
      }

      if (_this3.socketServer) {
        _logger.default.debug('Closing socket server.');

        yield _bluebird.default.promisify(_this3.socketServer.close, {
          context: _this3.socketServer
        })();
        _this3.socketServer = null;
      }
    })();
  }

  safeShutdown() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug('Shutting down command proxy and ignoring any errors');

      try {
        yield _this4.shutdown();
      } catch (err) {
        _logger.default.debug(`Ignoring error: ${err}`);
      }
    })();
  }

}

exports.UIAutoClient = UIAutoClient;
var _default = UIAutoClient;
exports.default = _default;require('source-map-support').install();
