// appium-ios-driver/build/lib/device-log/ios-crash-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSCrashLog = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _path = _interopRequireDefault(require("path"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSupport = require("appium-support");

var _logger = _interopRequireDefault(require("./logger"));

var _bluebird = _interopRequireDefault(require("bluebird"));

class IOSCrashLog {
  constructor(logDir) {
    this.logDir = logDir || _path.default.resolve(process.env.HOME || '/', 'Library', 'Logs', 'DiagnosticReports');
    this.prevLogs = [];
    this.logsSinceLastRequest = [];
  }

  getCrashes() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        return yield _appiumSupport.fs.glob(`${_this.logDir}/*.crash`);
      } catch (err) {
        _logger.default.errorAndThrow(`There was a problem getting the crash list: ${err}`);
      }
    })();
  }

  filesToJSON(files) {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return yield _bluebird.default.map(files, function () {
        var _ref = (0, _asyncToGenerator2.default)(function* (file) {
          let filename = _path.default.resolve(_this2.logDir, file);

          let stat = yield _appiumSupport.fs.stat(filename);
          return {
            timestamp: stat.ctime.getTime(),
            level: 'ALL',
            message: yield _appiumSupport.fs.readFile(filename, 'utf8')
          };
        });

        return function (_x) {
          return _ref.apply(this, arguments);
        };
      }());
    })();
  }

  startCapture() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this3.prevLogs = yield _this3.getCrashes();
    })();
  }

  stopCapture() {
    return (0, _asyncToGenerator2.default)(function* () {})();
  }

  getLogs() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let crashFiles = yield _this4.getCrashes();

      let diff = _lodash.default.difference(crashFiles, _this4.prevLogs, _this4.logsSinceLastRequest);

      _this4.logsSinceLastRequest = _lodash.default.union(_this4.logsSinceLastRequest, diff);
      return yield _this4.filesToJSON(diff);
    })();
  }

  getAllLogs() {
    var _this5 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let crashFiles = yield _this5.getCrashes();

      let logFiles = _lodash.default.difference(crashFiles, _this5.prevLogs);

      return yield _this5.filesToJSON(logFiles);
    })();
  }

}

exports.IOSCrashLog = IOSCrashLog;
var _default = IOSCrashLog;
exports.default = _default;require('source-map-support').install();
