// appium-ios-driver/build/lib/device-log/ios-performance-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSPerformanceLog = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

const MAX_EVENTS = 5000;

class IOSPerformanceLog {
  constructor(remoteDebugger, maxEvents = MAX_EVENTS) {
    this.remoteDebugger = remoteDebugger;
    this.maxEvents = parseInt(maxEvents, 10);
    this.timelineEvents = [];
  }

  startCapture() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug('Starting performance (Timeline) log capture');

      _this.timelineEvents = [];
      return yield _this.remoteDebugger.startTimeline(_this.onTimelineEvent.bind(_this));
    })();
  }

  stopCapture() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug('Stopping performance (Timeline) log capture');

      return yield _this2.remoteDebugger.stopTimeline();
    })();
  }

  onTimelineEvent(event) {
    _logger.default.debug(`Received Timeline event: ${_lodash.default.truncate(JSON.stringify(event))}`);

    this.timelineEvents.push(event);

    if (this.timelineEvents.length > this.maxEvents) {
      let removedEvent = this.timelineEvents.shift();

      _logger.default.warn(`Too many Timeline events, removing earliest: ${_lodash.default.truncate(JSON.stringify(removedEvent))}`);
    }
  }

  getLogs() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let events = _this3.timelineEvents;

      _logger.default.debug('Flushing Timeline events');

      _this3.timelineEvents = [];
      return events;
    })();
  }

  getAllLogs() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return _this4.getLogs();
    })();
  }

}

exports.IOSPerformanceLog = IOSPerformanceLog;
var _default = IOSPerformanceLog;
exports.default = _default;require('source-map-support').install();
