// appium-ios-driver/build/lib/cookies.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createJSCookie = createJSCookie;
exports.createJWPCookie = createJWPCookie;
exports.getValue = getValue;
exports.expireCookie = expireCookie;
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSupport = require("appium-support");

const log = _appiumSupport.logger.getLogger('Cookie');

function convertCookie(value, converter) {
  if (value.indexOf('"') === 0) {
    value = value.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
  }

  let parsedValue;

  try {
    parsedValue = decodeURIComponent(value.replace(/\+/g, ' '));
  } catch (e) {
    log.warn(e);
  }

  return converter ? converter(parsedValue) : parsedValue;
}

function createJSCookie(key, value, options = {}) {
  return [encodeURIComponent(key), '=', value, options.expires ? `; expires=${options.expires}` : '', options.path ? `; path=${options.path}` : '', options.domain ? `; domain=${options.domain}` : '', options.secure ? '; secure' : ''].join('');
}

function createJWPCookie(key, cookieString, converter = null) {
  let result = {};
  let cookies = cookieString ? cookieString.split('; ') : [];
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = cookies[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      let cookie = _step.value;
      let parts = cookie.split('=');
      let name = decodeURIComponent(parts.shift());
      let val = parts[0];

      if (key && key === name) {
        result.name = key;
        result.value = convertCookie(val, converter);
      } else {
        result[name] = convertCookie(val, converter);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return result;
}

function getValue(key, cookieString, converter = null) {
  let result = createJWPCookie(key, cookieString, converter);
  return _lodash.default.isUndefined(key) ? result : result.value;
}

function expireCookie(key, options) {
  return createJSCookie(key, '', _lodash.default.assign({}, options, {
    expires: 'Thu, 01 Jan 1970 00:00:00 GMT'
  }));
}

var _default = {
  createJSCookie,
  createJWPCookie,
  getValue,
  expireCookie
};
exports.default = _default;require('source-map-support').install();
