// appium-ios-driver/build/lib/iwdp.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IWDP = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _teen_process = require("teen_process");

var _events = _interopRequireDefault(require("events"));

var _logger = _interopRequireDefault(require("./logger"));

var _requestPromise = _interopRequireDefault(require("request-promise"));

var _asyncbox = require("asyncbox");

var _appiumSupport = require("appium-support");

const IWDP_CMD = 'ios_webkit_debug_proxy';
const MAX_RETRIES = 10;

let iwdpLogger = _appiumSupport.logger.getLogger('IWDP');

class IWDP extends _events.default {
  constructor(webkitDebugProxyPort = 27753, udid = null) {
    super();
    this.udid = udid;
    this.attempts = 0;
    this.port = webkitDebugProxyPort;
    this.process = this.createIWDPProcess();
    this.endpoint = `http://localhost:${this.port}`;
  }

  createIWDPProcess() {
    let process = new _teen_process.SubProcess(IWDP_CMD, ['-c', `${this.udid}:${this.port}`, '-d']);
    process.on('exit', () => this.onExit());
    process.on('lines-stderr', iwdpLogger.error);
    return process;
  }

  onExit() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (!_this.exitRequested) {
        _this.process = _this.createIWDPProcess();
        yield _this.start();
      }
    })();
  }

  start() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (++_this2.attempts > MAX_RETRIES) {
        return _this2.emit('error', new Error(`Failed to start IWDP server. Max retry attempts ${MAX_RETRIES} reached`));
      }

      _this2.exitRequested = false;

      if (!(yield _this2.isSupported())) {
        _logger.default.errorAndThrow(`'ios_webkit_debug_proxy' not installed on this machine. Try 'brew install ios-webkit-debug-proxy`);
      }

      _logger.default.debug(`Starting ios_webkit_debug_proxy at port ${_this2.port} on device ${_this2.udid}`);

      yield _this2.process.start(0);

      try {
        yield (0, _asyncbox.retryInterval)(20, 500, (0, _asyncToGenerator2.default)(function* () {
          return yield (0, _requestPromise.default)(_this2.endpoint);
        }));

        _this2.emit('start');
      } catch (ign) {
        try {
          yield _this2.process.stop();
        } catch (ign2) {}

        _logger.default.errorAndThrow(`Timed out waiting for ios_webkit_debug_proxy to open`);
      }
    })();
  }

  stop() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this3.attempts = 0;
      _this3.exitRequested = true;
      return yield _this3.process.stop();
    })();
  }

  isSupported() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (typeof _this4.supported !== 'undefined') {
        return _this4.supported;
      }

      try {
        yield _appiumSupport.fs.which(IWDP_CMD);
        _this4.supported = true;
      } catch (e) {
        _this4.supported = false;
      }

      return _this4.supported;
    })();
  }

}

exports.IWDP = IWDP;
var _default = IWDP;
exports.default = _default;require('source-map-support').install();
