// appium-ios-driver/build/lib/server.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startServer = startServer;
exports.startHttpsServer = startHttpsServer;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _logger = _interopRequireDefault(require("./logger"));

var _appiumBaseDriver = require("appium-base-driver");

var _driver = require("./driver");

var _https = _interopRequireDefault(require("https"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _requestPromise = _interopRequireDefault(require("request-promise"));

var _url = _interopRequireDefault(require("url"));

var _portfinder = _interopRequireDefault(require("portfinder"));

const pem = _bluebird.default.promisifyAll(require('pem'));

function startServer(_x, _x2) {
  return _startServer.apply(this, arguments);
}

function _startServer() {
  _startServer = (0, _asyncToGenerator2.default)(function* (port, host) {
    let driver = new _driver.IosDriver();
    let router = (0, _appiumBaseDriver.routeConfiguringFunction)(driver);
    let server = yield (0, _appiumBaseDriver.server)(router, port, host);

    _logger.default.info(`IosDriver server listening on http://${host}:${port}`);

    return server;
  });
  return _startServer.apply(this, arguments);
}

function startHttpsServer(_x3, _x4) {
  return _startHttpsServer.apply(this, arguments);
}

function _startHttpsServer() {
  _startHttpsServer = (0, _asyncToGenerator2.default)(function* (port, host) {
    let privateKey = yield pem.createPrivateKeyAsync();
    let keys = yield pem.createCertificateAsync({
      days: 1,
      selfSigned: true,
      serviceKey: privateKey.key
    });
    let pemCertificate = keys.certificate;
    let httpsPort = yield _portfinder.default.getPortPromise();
    const serverOpts = {
      key: keys.serviceKey,
      cert: pemCertificate
    };

    let sslServer = _https.default.createServer(serverOpts, function (req, res) {
      _logger.default.debug(`Received HTTPS '${req.method}' request for '${req.url}'`);

      if (req.method === 'OPTIONS') {
        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
        res.writeHead(200);
        res.end();
      } else if (req.method === 'POST') {
        let body = '';
        req.on('data', function (data) {
          body += data;
        });
        req.on('end', (0, _asyncToGenerator2.default)(function* () {
          let httpsUrl = _url.default.parse(req.url);

          let uri = `http://${host}:${port}${httpsUrl.pathname}`;

          _logger.default.debug(`Passing '${body}' to '${uri}'`);

          let options = {
            method: 'POST',
            uri,
            body
          };
          yield (0, _requestPromise.default)(options);
          res.writeHead(200);
          res.end();
        }));
      }
    }).listen(httpsPort);

    return {
      sslServer,
      pemCertificate,
      httpsPort
    };
  });
  return _startHttpsServer.apply(this, arguments);
}require('source-map-support').install();
