// appium-ios-driver/build/lib/commands/device.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;
commands.getDevicePixelRatio = (0, _asyncToGenerator2.default)(function* () {
  const screenHeight = (yield this.getWindowSize()).height;
  const screenHeightOfPlusDevices = 736;
  return screenHeight === screenHeightOfPlusDevices ? 3.0 : 2.0;
});
commands.getViewportRect = (0, _asyncToGenerator2.default)(function* () {
  const scale = yield this.getDevicePixelRatio();
  const statusBarHeight = Math.round((yield this.getStatusBarHeight()) * scale);
  const windowSize = yield this.getWindowSize();
  return {
    left: 0,
    top: statusBarHeight,
    width: windowSize.width * scale,
    height: windowSize.height * scale - statusBarHeight
  };
});
commands.getStatusBarHeight = (0, _asyncToGenerator2.default)(function* () {
  const command = 'UIATarget.localTarget().frontMostApp().statusBar().rect().size.height;';
  return yield this.uiAutoClient.sendCommand(command);
});
Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
