// appium-ios-driver/build/lib/commands/gesture.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _appiumBaseDriver = require("appium-base-driver");

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _utils = require("../utils");

var _logger = _interopRequireDefault(require("../logger"));

var _appiumSupport = require("appium-support");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;
const FLICK_MS = 3000;

commands.nativeTap = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (el) {
    el = (0, _utils.unwrapEl)(el);
    let command = `au.tapById('${el}')`;
    yield this.uiAutoClient.sendCommand(command);
  });

  return function (_x) {
    return _ref.apply(this, arguments);
  };
}();

commands.click = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (el) {
    el = (0, _utils.unwrapEl)(el);

    if (this.isWebContext()) {
      if (this.opts.nativeWebTap && !this.isRealDevice()) {
        yield this.nativeWebTap(el);
      } else {
        let atomsElement = this.useAtomsElement(el);
        return yield this.executeAtom('click', [atomsElement]);
      }
    } else {
      if (this.opts.useRobot) {
        throw new _appiumBaseDriver.errors.NotYetImplementedError();
      } else {
        yield this.nativeTap(el);
      }
    }
  });

  return function (_x2) {
    return _ref2.apply(this, arguments);
  };
}();

commands.clickCurrent = (0, _asyncToGenerator2.default)(function* () {
  if (this.isWebContext()) {
    if (_lodash.default.isNull(this.curWebCoords)) {
      throw new _appiumBaseDriver.errors.UnknownError('Cannot call click() before calling moveTo() to set coords');
    }

    yield this.clickWebCoords();
  } else {
    if (this.curCoords === null) {
      throw new _appiumBaseDriver.errors.UnknownError("Cannot call click() before calling moveTo() to set coords");
    }

    yield this.clickCoords(this.curCoords);
  }
});

helpers.clickCoords = function () {
  var _ref4 = (0, _asyncToGenerator2.default)(function* (coords) {
    if (this.opts.useRobot) {
      throw new _appiumBaseDriver.errors.NotYetImplementedError();
    } else {
      let opts = coords;
      opts.tapCount = 1;
      opts.duration = 0.3;
      opts.touchCount = 1;
      let command = `au.complexTap(${JSON.stringify(opts)})`;
      yield this.uiAutoClient.sendCommand(command);
    }
  });

  return function (_x3) {
    return _ref4.apply(this, arguments);
  };
}();

commands.mobileRotation = function () {
  var _ref5 = (0, _asyncToGenerator2.default)(function* (x, y, radius, rotation, touchCount, duration, el) {
    if (this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotYetImplementedError();
    }

    el = (0, _utils.unwrapEl)(el);
    let location = {
      x,
      y
    };
    let options = {
      duration,
      radius,
      rotation,
      touchCount
    };

    if (el) {
      if (this.isWebContext()) {
        throw new _appiumBaseDriver.errors.NotYetImplementedError();
      }

      yield this.uiAutoClient.sendCommand(`au.getElement('${el}').rotateWithOptions(${JSON.stringify(options)})`);
    } else {
      yield this.uiAutoClient.sendCommand(`target.rotateWithOptions(${JSON.stringify(location)}, ${JSON.stringify(options)})`);
    }
  });

  return function (_x4, _x5, _x6, _x7, _x8, _x9, _x10) {
    return _ref5.apply(this, arguments);
  };
}();

helpers.handleTap = function () {
  var _ref6 = (0, _asyncToGenerator2.default)(function* (gesture) {
    let options = gesture.options;
    let cmdBase = options.element ? `au.getElement('${options.element}')` : 'UIATarget.localTarget().frontMostApp()';
    let rect = yield this.uiAutoClient.sendCommand(`${cmdBase}.rect()`);
    let offsetX = 0.5;
    let offsetY = 0.5;
    let size = {
      w: rect.size.width,
      h: rect.size.height
    };
    options.x = options.x || size.w / 2;
    options.y = options.y || size.h / 2;
    offsetX = options.x / size.w;
    offsetY = options.y / size.h;
    let opts = {
      tapOffset: {
        x: offsetX,
        y: offsetY
      },
      tapCount: options.count || 1,
      touchCount: 1
    };
    let cmd = `${cmdBase}.tapWithOptions(${JSON.stringify(opts)})`;
    return yield this.uiAutoClient.sendCommand(cmd);
  });

  return function (_x11) {
    return _ref6.apply(this, arguments);
  };
}();

helpers.isDrag = function (gestures) {
  return gestures.length === 4 && gestures[0].action === 'press' && gestures[1].action === 'wait' && gestures[2].action === 'moveTo' && gestures[3].action === 'release';
};

helpers.isPinchAndZoom = function (gestures) {
  return gestures.length === 3 && gestures[0].action === 'press' && gestures[1].action === 'moveTo' && gestures[2].action === 'release';
};

helpers.getCoordinates = function () {
  var _ref7 = (0, _asyncToGenerator2.default)(function* (gesture) {
    let el = gesture.options.element;
    let coordinates = {
      x: 0,
      y: 0,
      areOffsets: false
    };

    if (el) {
      let command = `au.getElement('${el}').rect()`;
      let rect = yield this.uiAutoClient.sendCommand(command);
      let pos = {
        x: rect.origin.x,
        y: rect.origin.y
      };
      let size = {
        w: rect.size.width,
        h: rect.size.height
      };
      let offsetX = 0;
      let offsetY = 0;

      if (gesture.options.x || gesture.options.y) {
        offsetX = gesture.options.x || 0;
        offsetY = gesture.options.y || 0;
      } else {
        offsetX = size.w / 2;
        offsetY = size.h / 2;
      }

      coordinates.x = pos.x + offsetX;
      coordinates.y = pos.y + offsetY;
    } else {
      coordinates.areOffsets = gesture.action === 'moveTo';
      coordinates.x = gesture.options.x || 0;
      coordinates.y = gesture.options.y || 0;
    }

    return coordinates;
  });

  return function (_x12) {
    return _ref7.apply(this, arguments);
  };
}();

helpers.applyMoveToOffset = function (firstCoordinates, secondCoordinates) {
  if (secondCoordinates.areOffsets) {
    return {
      x: firstCoordinates.x + secondCoordinates.x,
      y: firstCoordinates.y + secondCoordinates.y
    };
  } else {
    return secondCoordinates;
  }
};

helpers.handleDrag = function () {
  var _ref8 = (0, _asyncToGenerator2.default)(function* (gestures) {
    let press = gestures[0];
    let wait = gestures[1];
    let moveTo = gestures[2];
    let pressCoordinates = yield this.getCoordinates(press);
    let duration = parseInt(wait.options.ms, 10) / 1000;
    let moveToCoordinates = yield this.getCoordinates(moveTo);
    moveToCoordinates = this.applyMoveToOffset(pressCoordinates, moveToCoordinates);
    let dragCommand = `au.dragApp(` + `${pressCoordinates.x}, ${pressCoordinates.y}, ` + `${moveToCoordinates.x}, ${moveToCoordinates.y}, ` + `${duration})`;
    return yield this.uiAutoClient.sendCommand(dragCommand);
  });

  return function (_x13) {
    return _ref8.apply(this, arguments);
  };
}();

helpers.handlePinchAndZoom = function () {
  var _ref9 = (0, _asyncToGenerator2.default)(function* (gestures) {
    let press = gestures[0];
    let moveTo = gestures[1];
    let pressCoordinates = yield this.getCoordinates(press);
    let moveToCoordinates = yield this.getCoordinates(moveTo);
    moveToCoordinates = this.applyMoveToOffset(pressCoordinates, moveToCoordinates);
    let pinchCommand = `au.dragApp(` + `${pressCoordinates.x}, ${pressCoordinates.y}, ` + `${moveToCoordinates.x}, ${moveToCoordinates.y}`;
    return yield this.uiAutoClient.sendCommand(pinchCommand);
  });

  return function (_x14) {
    return _ref9.apply(this, arguments);
  };
}();

commands.performTouch = function () {
  var _ref10 = (0, _asyncToGenerator2.default)(function* (gestures) {
    if (this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotYetImplementedError();
    }

    if (gestures.length === 1 && gestures[0].action === 'tap') {
      return yield this.handleTap(gestures[0]);
    } else if (this.isDrag(gestures)) {
      return yield this.handleDrag(gestures);
    }

    let touchStateObjects = yield this.parseTouch(gestures);
    yield this.uiAutoClient.sendCommand(`target.touch(${JSON.stringify(touchStateObjects)})`);
  });

  return function (_x15) {
    return _ref10.apply(this, arguments);
  };
}();

commands.parseTouch = function () {
  var _ref11 = (0, _asyncToGenerator2.default)(function* (gestures) {
    var _this = this;

    if (_lodash.default.last(gestures).action === 'release') {
      gestures.pop();
    }

    let touchStateObjects = [];

    let finishParsing = () => {
      let prevPos = null;
      let time = 0;

      for (var _i = 0; _i < touchStateObjects.length; _i++) {
        let state = touchStateObjects[_i];

        if (state.touch[0] === false) {
          state.touch[0] = prevPos;
        } else if (state.touch[0].offset && prevPos) {
          state.touch[0].x += prevPos.x;
          state.touch[0].y += prevPos.y;
        }

        if (state.touch[0]) {
          delete state.touch[0].offset;
          prevPos = state.touch[0];
        }

        let timeOffset = state.timeOffset;
        time += timeOffset;
        state.time = time;
        delete state.timeOffset;
      }
    };

    let needsPoint = action => {
      return _lodash.default.includes(['press', 'moveTo', 'tap', 'longPress'], action);
    };

    let cycleThroughGestures = function () {
      var _ref12 = (0, _asyncToGenerator2.default)(function* () {
        let gesture = gestures.shift();

        if (typeof gesture === "undefined") {
          return finishParsing();
        }

        let tapPoint = false;

        if (needsPoint(gesture.action)) {
          let el = gesture.options.element;

          if (el) {
            let command = `au.getElement('${el}').rect()`;
            let rect = yield _this.uiAutoClient.sendCommand(command);
            let pos = {
              x: rect.origin.x,
              y: rect.origin.y
            };
            let size = {
              w: rect.size.width,
              h: rect.size.height
            };

            if (gesture.options.x || gesture.options.y) {
              tapPoint = {
                offset: false,
                x: pos.x + (gesture.options.x || 0),
                y: pos.y + (gesture.options.y || 0)
              };
            } else {
              tapPoint = {
                offset: false,
                x: pos.x + size.w / 2,
                y: pos.y + size.h / 2
              };
            }

            let touchStateObject = {
              timeOffset: 0.2,
              touch: [tapPoint]
            };
            touchStateObjects.push(touchStateObject);
            yield cycleThroughGestures();
          } else {
            tapPoint = {
              offset: true,
              x: gesture.options.x || 0,
              y: gesture.options.y || 0
            };
            let touchStateObject = {
              timeOffset: 0.2,
              touch: [tapPoint]
            };
            touchStateObjects.push(touchStateObject);
            yield cycleThroughGestures();
          }
        } else {
          tapPoint = false;
          let offset = 0.2;

          if (gesture.action === 'wait') {
            if (typeof gesture.options.ms !== 'undefined' || gesture.options.ms !== null) {
              offset = parseInt(gesture.options.ms, 10) / 1000;
            }
          }

          let touchStateObject = {
            timeOffset: offset,
            touch: [tapPoint]
          };
          touchStateObjects.push(touchStateObject);
          yield cycleThroughGestures();
        }
      });

      return function cycleThroughGestures() {
        return _ref12.apply(this, arguments);
      };
    }();

    yield cycleThroughGestures();
    return touchStateObjects;
  });

  return function (_x16) {
    return _ref11.apply(this, arguments);
  };
}();

let mergeStates = function mergeStates(states) {
  let getSlice = function getSlice(states, index) {
    let array = [];

    for (let i = 0; i < states.length; i++) {
      array.push(states[i][index]);
    }

    return array;
  };

  let timeSequence = function timeSequence(states) {
    let seq = [];

    _lodash.default.each(states, function (state) {
      let times = _lodash.default.map(state, "time");

      seq = _lodash.default.union(seq, times);
    });

    return seq.sort();
  };

  let merged = [];

  _lodash.default.each(timeSequence(states), function (time, index) {
    let slice = getSlice(states, index);
    let obj = {
      time,
      touch: []
    };

    _lodash.default.each(slice, function (action) {
      obj.touch.push(action.touch[0]);
    });

    merged.push(obj);
  });

  return merged;
};

commands.performMultiAction = function () {
  var _ref13 = (0, _asyncToGenerator2.default)(function* (actions) {
    var _this2 = this;

    if (this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotYetImplementedError();
    }

    let states = [];

    let cycleThroughActions = function () {
      var _ref14 = (0, _asyncToGenerator2.default)(function* () {
        let action = actions.shift();

        if (typeof action === "undefined") {
          let mergedStates = mergeStates(states);

          if (_this2.isPinchAndZoom(mergedStates)) {
            yield _this2.handlePinchAndZoom(mergedStates);
            return;
          } else {
            yield _this2.uiAutoClient.sendCommand(`target.touch(${JSON.stringify(mergedStates)})`);
            return;
          }
        }

        let val = yield _this2.parseTouch(action);
        states.push(val);
        yield cycleThroughActions();
      });

      return function cycleThroughActions() {
        return _ref14.apply(this, arguments);
      };
    }();

    yield cycleThroughActions();
  });

  return function (_x17) {
    return _ref13.apply(this, arguments);
  };
}();

helpers.mobileScroll = function () {
  var _ref15 = (0, _asyncToGenerator2.default)(function* (opts = {}) {
    let direction = opts.direction;
    let el = opts.element;
    el = (0, _utils.unwrapEl)(el);

    if (this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotYetImplementedError();
    } else {
      if (_appiumSupport.util.hasValue(el) && !_appiumSupport.util.hasValue(direction)) {
        yield this.scrollToElement(el);
        return;
      }

      direction = _lodash.default.capitalize(direction);
      let command;

      if (_lodash.default.isNull(el) || _lodash.default.isUndefined(el)) {
        command = `au.scrollFirstView('${direction}')`;
      } else {
        command = `au.getElement('${el}').scroll${direction}()`;
      }

      try {
        yield this.uiAutoClient.sendCommand(command);
      } catch (err) {
        if (!_lodash.default.includes(err.message, 'kAXErrorFailure')) throw err;

        _logger.default.warn('Received kAXErrorFailure, generally indicating an attempt ' + 'to scroll past the end of the view. Continuing.');
      }
    }
  });

  return function () {
    return _ref15.apply(this, arguments);
  };
}();

commands.scrollToElement = function () {
  var _ref16 = (0, _asyncToGenerator2.default)(function* (element) {
    element = (0, _utils.unwrapEl)(element);
    const command = `au.getElement('${element}').scrollToVisible()`;
    yield this.uiAutoClient.sendCommand(command);
  });

  return function (_x18) {
    return _ref16.apply(this, arguments);
  };
}();

commands.flick = function () {
  var _ref17 = (0, _asyncToGenerator2.default)(function* (el, xspeed, yspeed, xoffset, yoffset, speed) {
    el = (0, _utils.unwrapEl)(el);

    if (_lodash.default.every([el, xoffset, yoffset, speed], p => {
      return !_lodash.default.isNull(p) && !_lodash.default.isUndefined(p);
    })) {
      yield this.flickElement(el, xoffset, yoffset, speed);
    } else if (_lodash.default.every([xspeed, yspeed], p => {
      return !_lodash.default.isNull(p) && !_lodash.default.isUndefined(p);
    })) {
      yield this.xySpeedFlick(xspeed, yspeed);
    } else {
      throw new _appiumBaseDriver.errors.UnknownError('Bad flick parameters, pass either (xspeed, yspeed) or (element, xoffset, yoffset, speed)!');
    }
  });

  return function (_x19, _x20, _x21, _x22, _x23, _x24) {
    return _ref17.apply(this, arguments);
  };
}();

helpers.xySpeedFlick = function () {
  var _ref18 = (0, _asyncToGenerator2.default)(function* (xSpeed, ySpeed) {
    let command = `au.touchFlickFromSpeed(${xSpeed},${ySpeed})`;
    yield _bluebird.default.all([this.uiAutoClient.sendCommand(command), _bluebird.default.delay(FLICK_MS)]);
  });

  return function (_x25, _x26) {
    return _ref18.apply(this, arguments);
  };
}();

helpers.flickElement = function () {
  var _ref19 = (0, _asyncToGenerator2.default)(function* (el, xoffset, yoffset, speed) {
    el = (0, _utils.unwrapEl)(el);
    let command = "";

    if (this.isWebContext()) {
      yield this.webFlickElement(el, xoffset, yoffset);
    } else {
      command = `au.getElement('${el}').touchFlick(${xoffset},${yoffset},${speed})`;
      yield _bluebird.default.all([this.uiAutoClient.sendCommand(command), _bluebird.default.delay(FLICK_MS)]);
    }
  });

  return function (_x27, _x28, _x29, _x30) {
    return _ref19.apply(this, arguments);
  };
}();

commands.mobileShake = (0, _asyncToGenerator2.default)(function* () {
  yield this.uiAutoClient.sendCommand("au.shake()");
});

commands.moveTo = function () {
  var _ref21 = (0, _asyncToGenerator2.default)(function* (el, xoffset = 0, yoffset = 0) {
    el = (0, _utils.unwrapEl)(el);

    if (this.isWebContext()) {
      let _ref22 = yield this.getLocation(el),
          x = _ref22.x,
          y = _ref22.y;

      let coords = {
        x: x + xoffset,
        y: y + yoffset
      };
      this.curWebCoords = coords;
      let atomsElement = this.useAtomsElement(el);
      let relCoords = {
        x: xoffset,
        y: yoffset
      };
      yield this.executeAtom('move_mouse', [atomsElement, relCoords]);
    } else {
      if (_lodash.default.isNull(el) || _lodash.default.isUndefined(el)) {
        if (!this.curCoords) {
          throw new _appiumBaseDriver.errors.UnknownException('Current cursor position unknown, please use moveTo with an element the first time.');
        }

        this.curCoords = {
          x: this.curCoords.x + xoffset,
          y: this.curCoords.y + yoffset
        };
      } else {
        let elPos = yield this.getLocation(el);
        this.curCoords = {
          x: elPos.x + xoffset,
          y: elPos.y + yoffset
        };
      }
    }
  });

  return function (_x31) {
    return _ref21.apply(this, arguments);
  };
}();

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
