// appium-ios-driver/build/lib/commands/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.iosCommands = exports.commands = exports.default = void 0;

var _find = _interopRequireDefault(require("./find"));

var _element = _interopRequireDefault(require("./element"));

var _logging = _interopRequireDefault(require("./logging"));

var _gesture = _interopRequireDefault(require("./gesture"));

var _alert = _interopRequireDefault(require("./alert"));

var _execute = _interopRequireDefault(require("./execute"));

var _general = _interopRequireDefault(require("./general"));

var _context = _interopRequireDefault(require("./context"));

var _web = _interopRequireDefault(require("./web"));

var _orientation = _interopRequireDefault(require("./orientation"));

var _fileMovement = _interopRequireDefault(require("./file-movement"));

var _navigation = _interopRequireDefault(require("./navigation"));

var _screenshot = _interopRequireDefault(require("./screenshot"));

var _device = _interopRequireDefault(require("./device"));

var _safari = _interopRequireDefault(require("./safari"));

var _coverage = _interopRequireDefault(require("./coverage"));

var _timeout = _interopRequireDefault(require("./timeout"));

let commands = {};
exports.commands = commands;
var _arr = [_find.default, _element.default, _logging.default, _gesture.default, _alert.default, _execute.default, _general.default, _context.default, _web.default, _orientation.default, _fileMovement.default, _navigation.default, _screenshot.default, _device.default, _safari.default, _coverage.default, _timeout.default];

for (var _i = 0; _i < _arr.length; _i++) {
  let obj = _arr[_i];
  Object.assign(commands, obj);
}

let iosCommands = {
  find: _find.default,
  element: _element.default,
  logging: _logging.default,
  gesture: _gesture.default,
  alert: _alert.default,
  execute: _execute.default,
  general: _general.default,
  context: _context.default,
  web: _web.default,
  orientation: _orientation.default,
  file: _fileMovement.default,
  navigation: _navigation.default,
  screenshot: _screenshot.default,
  device: _device.default,
  safari: _safari.default,
  coverage: _coverage.default,
  timeout: _timeout.default
};
exports.iosCommands = iosCommands;
var _default = commands;
exports.default = _default;require('source-map-support').install();
