// appium-ios-driver/build/lib/commands/screenshot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _uuidJs = _interopRequireDefault(require("uuid-js"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _path = _interopRequireDefault(require("path"));

var _asyncbox = require("asyncbox");

var _appiumSupport = require("appium-support");

var _utils = require("../uiauto/utils");

var _logger = _interopRequireDefault(require("../logger"));

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;
commands.getScreenshot = (0, _asyncToGenerator2.default)(function* () {
  var _this = this;

  let guid = _uuidJs.default.create();

  let shotFile = `screenshot${guid}`;

  let shotFolder = _path.default.resolve(this.opts.tmpDir, 'appium-instruments/Run 1/');

  if (!(yield _appiumSupport.fs.exists(shotFolder))) {
    _logger.default.debug(`Creating folder '${shotFolder}'`);

    yield (0, _appiumSupport.mkdirp)(shotFolder);
  }

  let shotPath = _path.default.resolve(shotFolder, `${shotFile}.png`);

  _logger.default.debug(`Taking screenshot: '${shotPath}'`);

  let takeScreenShot = function () {
    var _ref2 = (0, _asyncToGenerator2.default)(function* () {
      yield _this.uiAutoClient.sendCommand(`au.capture('${shotFile}')`);
      let screenshotWaitTimeout = (_this.opts.screenshotWaitTimeout || 10) * 1000;

      _logger.default.debug(`Waiting ${screenshotWaitTimeout} ms for screenshot to be generated.`);

      let startMs = Date.now();
      let success = false;

      while (Date.now() - startMs < screenshotWaitTimeout) {
        if (yield _appiumSupport.fs.hasAccess(shotPath)) {
          success = true;
          break;
        }

        yield _bluebird.default.delay(300);
      }

      if (!success) {
        throw new _appiumBaseDriver.errors.UnknownError('Timed out waiting for screenshot file');
      }

      if ((yield _this.getOrientation()) === 'LANDSCAPE') {
        _logger.default.debug('Rotating landscape screenshot');

        yield (0, _utils.rotateImage)(shotPath, -90);
      }

      return yield _appiumSupport.fs.readFile(shotPath);
    });

    return function takeScreenShot() {
      return _ref2.apply(this, arguments);
    };
  }();

  let data = yield (0, _asyncbox.retry)(3, takeScreenShot);
  return Buffer.from(data).toString('base64');
});
commands.getViewportScreenshot = (0, _asyncToGenerator2.default)(function* () {
  const windowSize = yield this.getWindowSize();
  const scale = yield this.getDevicePixelRatio();
  const statusBarHeight = (yield this.getStatusBarHeight()) * scale;
  const screenshot = yield this.getScreenshot();
  let rect = {
    left: 0,
    top: statusBarHeight,
    width: windowSize.width * scale,
    height: windowSize.height * scale - statusBarHeight
  };
  let newScreenshot = yield _appiumSupport.imageUtil.cropBase64Image(screenshot, rect);
  return newScreenshot;
});
Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
